zips/draft-stark-verify-nu.md

```diff
@@ -0,0 +1,150 @@
+* Define a `param_id` namespace that pins the specification of the Stwo proof relation, the proof object wire format, field encodings, transcript/Merkle hash suite and personalization, FRI expansion and query caps, any grinding bounds, and any batch‑verification settings.
+* A reference implementation of the prover and verifier, and reference test vectors (valid and invalid) must be provided for each such `param_id`.
+* Limits on proof sizes, etc. must be chosen to avoid potential denial of service, and the necessary updates to ZIP 317 fee calculation [^zip-0317] must be specified.
+
+## Specification
+
+### TZE identification
+
+- `tze_id`: TBD (assigned by ZIP Editors).
+- `tze_mode`: 0 (only one mode is defined by this ZIP). Any other value MUST be rejected. 
+
+All serialization below is inside the `tze_data_payload` defined by [^zip-0222]; lengths are `compactSize`.
+
+### Precondition (prefix‑free)
+
+```
+u16  param_id         // selects a pinned Stwo verifier parameter set (see below)
+u8   flags            // bit 0 = bind_ctx; all other bits MUST be 0
+vec  pub_in           // public inputs to the proved statement (len + bytes)
+if (flags & 0x01) {
+    bytes32 ctx_digest   // non-malleable transaction digest
+}
+```
+
+- `param_id` selects an enumerated, pinned Stwo parameter set (field & hash suite, FRI/query caps, wire format version, endianness, personalization, etc.). Unknown `param_id` MUST be rejected.
+- `pub_in` is an opaque byte string to consensus; it MUST be absorbed by the verifier transcript exactly as specified (byte‑for‑byte).
+- If `bind_ctx` is set, `ctx_digest` MUST equal the transaction's non‑malleable digest [^zip-0244]. A mismatch MUST cause rejection.
+
+### Witness (prefix‑free)
+
+```
+vec  proof            // the Stwo (Circle STARK) proof object (len + bytes)
+```
+
+The `proof` MUST conform to the Stwo wire format pinned by `param_id`.
+
+### Verification
+
+`bool tze_verify(0, precondition, witness, context)`:
+
+1. Parse `param_id`; load the exact pinned parameter set. Reject if unknown.
+2. Parse `flags`; if any reserved bits set, reject.
+3. If `bind_ctx`, compute the required `ctx_digest` from `context` and compare; on mismatch, reject. ([Zcash Zips][3])
+4. Parse `pub_in` and `proof`. Enforce all consensus bounds (see Bounds).
+5. Initialize the Stwo transcript/channel, **absorb `pub_in` exactly**, and verify `proof` using the parameters pinned by `param_id`.
+6. Return `true` iff the verifier accepts; otherwise return `false`.
+
+### Parameter sets (`param_id`)
+
+The namespace is 16‑bit (0x0001..0xFFFF). This ZIP introduces:
+
+- `param_id = 0x0001` — `STWO_V1_P1`
+
+  - Upstream: `starkware-libs/stwo` at tag TBD, commit `TBD`.
+  - Wire format: Stwo proof object at v1.0.0 (exact byte layout pinned in this ZIP's test‑vector directory).
```

Comment:
We don't normally rely on test vectors as part of a specification. Rather, test vectors are supposed to be a redundant check that can be compared against both the specification and an implementation. (Ditto for the four lines below.)